### PR TITLE
Clang Static Analyzer

### PR DESCRIPTION
**Pull request by lpascal-ledger**

### CODE REVIEW

**Comment by lpascal-ledger** [comment]
Also added some Tezos test advising the application as "Tezos" instead of "Tezos Wallet", as a function with a potentially shady behavior converts "Tezos" to "Tezos Wallet".

File: .github/workflows/ci-workflow.yml [hunk 1]
```diff
@@ -200,3 +200,20 @@ jobs:
 
       - name: Run zemu tests
         run: cd test && yarn test -t "\[Nano SP\]"
+
+  job_scan_build:
```
Comment (fbeutin-ledger) [inline .github/workflows/ci-workflow.yml:204]: So this is NanoS only scan ?

File: src/parse_coin_config.c [hunk 2]
```diff
@@ -15,10 +15,13 @@ void set_ledger_application_name(buf_t *application_name) {
     for (size_t i = 0; i < sizeof(appnames_aliases) / sizeof(appnames_aliases[0]); i++) {
         if (!strcmp((const char *) application_name->bytes,
                     (char *) (PIC(appnames_aliases[i].foreign_name)))) {
+            const size_t app_name_size = strlen((char *) PIC(appnames_aliases[i].app_name));
             // Copy the correct application name.
-            strcpy((char *) application_name->bytes, (char *) (PIC(appnames_aliases[i].app_name)));
+            strlcpy((char *) application_name->bytes,
```
Comment (fbeutin-ledger) [inline src/parse_coin_config.c:20]: Maybe we could check the return of strlcpy

File: src/parse_coin_config.c [hunk 3]
```diff
@@ -15,10 +15,13 @@ void set_ledger_application_name(buf_t *application_name) {
...
+                    (char *) (PIC(appnames_aliases[i].app_name)),
+                    app_name_size + 1);
```
Comment (fbeutin-ledger) [inline src/parse_coin_config.c:22]: I think this field should be the size of dst == the max size to copy

File: src/parse_coin_config.c [hunk 4]
```diff
@@ -15,10 +15,13 @@ void set_ledger_application_name(buf_t *application_name) {
...
             // Update the application name size accordingly.
-            application_name->size = strlen((char *) PIC(appnames_aliases[i].app_name));
+            application_name->size = app_name_size;
```
Comment (fbeutin-ledger) [inline src/parse_coin_config.c:24]: Only true if strlcpy succeded with no truncation

File: src/process_transaction.c [hunk 5]
```diff
@@ -27,7 +27,9 @@ void to_uppercase(char *str, unsigned char size) {
 void set_ledger_currency_name(char *currency) {
     for (size_t i = 0; i < sizeof(currencies_aliases) / sizeof(currencies_aliases[0]); i++) {
         if (!strcmp(currency, (char *) (PIC(currencies_aliases[i].foreign_name)))) {
-            strcpy(currency, (char *) (PIC(currencies_aliases[i].ledger_name)));
+            strlcpy(currency,
+                    (char *) (PIC(currencies_aliases[i].ledger_name)),
+                    strlen((char *) PIC(currencies_aliases[i].ledger_name)));
```
Comment (fbeutin-ledger) [inline src/process_transaction.c:32]: this field should be the max size currency can be written into before overflowing

File: test/src/tezos.test.js [hunk 6]
```diff
@@ -4,56 +4,68 @@ import "./protocol_pb.js";
...
-        t.setToCurrencyInfo(XTZ_INFO);
-        t.setAmountToProvider(100000000);
-        t.setAmountToWallet(1000000000);
-        t.setFee(10000000);
-        await t.performInvalidPayoutTransaction();
-    }))
+const XTZ_variants = [{'name': 'normal', 'config': XTZ_INFO},
+                      {'name': 'legacy', 'config': XTZ_LEGACY_INFO}];
+
+XTZ_variants.forEach(function(variant) {
```
Comment (fbeutin-ledger) [inline test/src/tezos.test.js:17]: Good idea, we could use it with other coins in time

File: .github/workflows/ci-workflow.yml (hunk 1 above)
Comment (lpascal-ledger) [reply .github/workflows/ci-workflow.yml:204]: It generally is. In this case, the only non-NanoS code is 
```
#ifdef TARGET_NANOX
                // grab the current plane mode setting
                G_io_app.plane_mode = os_setting_get(OS_SETTING_PLANEMODE, NULL, 0);
#endif  // TARGET_NANOX
```

File: src/parse_coin_config.c (hunk 2 above)
Comment (lpascal-ledger) [reply src/parse_coin_config.c:20]: I'll rewrite this function, the original logic seems flawed.

File: src/process_transaction.c (hunk 5 above)
Comment (lpascal-ledger) [reply src/process_transaction.c:32]: Actually if think it should be something like `MIN(sizeof(<array>currency), strlen((char *) PIC(currencies_aliases[i].ledger_name)))`.
This size is [purely local](https://github.com/LedgerHQ/app-exchange/blob/develop/src/proto/protocol.pb.h#L40-L41) for now, I'll make it more public.

**Review by fbeutin-ledger — APPROVED** [review]

File: src/parse_coin_config.c [hunk 7]
```diff
@@ -1,6 +1,7 @@
+#include <string.h>
+#include <os.h>
+
 #include "parse_coin_config.h"
-#include "string.h"
-#include "os.h"
 
 typedef struct app_name_alias_s {
     char *foreign_name;
```
Comment (fbeutin-ledger) [inline src/parse_coin_config.c:7]: const

Comment (fbeutin-ledger) [reply src/parse_coin_config.c:7]: This comment applies to the fields of app_name_alias_s, github somehow placed it here

File: src/process_transaction.c [hunk 8]
```diff
@@ -24,10 +24,14 @@ void to_uppercase(char *str, unsigned char size) {
...
+void set_ledger_currency_name(char *currency, size_t currency_size) {
     for (size_t i = 0; i < sizeof(currencies_aliases) / sizeof(currencies_aliases[0]); i++) {
-        if (!strcmp(currency, (char *) (PIC(currencies_aliases[i].foreign_name)))) {
-            strcpy(currency, (char *) (PIC(currencies_aliases[i].ledger_name)));
+        if (!strncmp(currency,
+                     (char *) (PIC(currencies_aliases[i].foreign_name)),
+                     strlen((char *) PIC(currencies_aliases[i].foreign_name)))) {
+            strlcpy(currency,
+                    (char *) (PIC(currencies_aliases[i].ledger_name)),
+                    MIN(strlen((char *) PIC(currencies_aliases[i].ledger_name)), currency_size));
```
[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/process_transaction.c:34]: Indeed, replacing with `currency_size` only.

File: src/process_transaction.c [hunk 9]
```diff
@@ -65,8 +69,12 @@ void normalize_currencies(swap_app_context_t *ctx) {
                  sizeof(ctx->received_transaction.currency_from));
     to_uppercase(ctx->received_transaction.currency_to,
                  sizeof(ctx->received_transaction.currency_to));
-    set_ledger_currency_name(ctx->received_transaction.currency_from);
-    set_ledger_currency_name(ctx->received_transaction.currency_to);
+    set_ledger_currency_name(ctx->received_transaction.currency_from,
+                             sizeof(ctx->received_transaction.currency_from) /
+                                 sizeof(ctx->received_transaction.currency_from[0]));
```
[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/process_transaction.c:74]: The type of `ctx->received_transaction.currency_from` is unknown from this code point of view. Should it become an int or any other type, this code would have to be change. I prefer keeping it that way, even if I know it is a division by 1.

File: src/parse_coin_config.c [hunk 10]
```diff
@@ -11,50 +12,52 @@ const app_name_alias_t appnames_aliases[] = {
...
     if (config.size != 3 + ticker->size + application_name->size + pure_config->size) return 0;
     if (pure_config->size > 0)
         pure_config->bytes = config.bytes + 1 + ticker->size + 1 + application_name->size + 1;
 
     // Update the application name to match Ledger's naming.
-    set_ledger_application_name(application_name);
+    for (size_t i = 0; i < sizeof(appnames_aliases) / sizeof(appnames_aliases[0]); i++) {
+        if (!strncmp((const char *) application_name->bytes,
+                     (char *) (PIC(appnames_aliases[i].foreign_name)),
+                     strlen((char *) PIC(appnames_aliases[i].foreign_name)))) {
```
[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/parse_coin_config.c:55]: `sizeof` won't do, as variables are pointers, there size would always be 4 whatever the string. 
However, as we know the size of a (`application_name->size`), I will check for
`MIN(application_name->size, strlen((char *) PIC(appnames_aliases[i].foreign_name)))`

[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/parse_coin_config.c]: I think the `strlen()` is redundant with `strncmp()`, that already stops upon encountering a `\0` char.

**Pull request merged**